4.0.1版本 installHook会报异常

更新到最新4.0.1版本，应用补丁重启后会看到如下异常：

HookFactory: installHook me.ele.amigo.hook.IActivityManagerHook@c58ef4a errorandroid.util.AndroidRuntimeException: Can not install IActivityManagerNative hook
                                                                 at me.ele.amigo.hook.IActivityManagerHook.onInstall(IActivityManagerHook.java:87)
                                                                 at me.ele.amigo.hook.HookFactory.installHook(HookFactory.java:23)
                                                                 at me.ele.amigo.hook.HookFactory.install(HookFactory.java:17)
                                                                 at java.lang.reflect.Method.invoke(Native Method)
                                                                 at me.ele.amigo.reflect.MethodUtils.invokeStaticMethod(MethodUtils.java:33)
                                                                 at me.ele.amigo.reflect.MethodUtils.invokeStaticMethod(MethodUtils.java:40)
                                                                 at me.ele.amigo.Amigo.installHook(Amigo.java:188)
                                                                 at me.ele.amigo.Amigo.runPatchApk(Amigo.java:158)
                                                                 at me.ele.amigo.Amigo.onCreate(Amigo.java:112)
                                                                 at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1013)
                                                                 at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4707)
                                                                 at android.app.ActivityThread.-wrap1(ActivityThread.java)
                                                                 at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1405)
                                                                 at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                 at android.os.Looper.loop(Looper.java:148)
                                                                 at android.app.ActivityThread.main(ActivityThread.java:5417)
                                                                 at java.lang.reflect.Method.invoke(Native Method)
                                                                 at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
                                                                 at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)

@peacepassion 把Amigo demo使用混淆，发现异常，日志如下：

0-13 16:17:05.315 27890-27890/me.ele.app.amigo W/System.err: java.lang.ClassNotFoundException: Didn't find class "me.ele.app.amigo.receiver.DemoReceiver" on path: DexPathList[[zip file "/data/app/me.ele.app.amigo-1/base.apk"],nativeLibraryDirectories=[/data/app/me.ele.app.amigo-1/lib/arm64, /system/lib64, /vendor/lib64]]
10-13 16:17:05.316 27890-27890/me.ele.app.amigo W/System.err:     at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
10-13 16:17:05.316 27890-27890/me.ele.app.amigo W/System.err:     at java.lang.ClassLoader.loadClass(ClassLoader.java:380)
10-13 16:17:05.316 27890-27890/me.ele.app.amigo W/System.err:     at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
10-13 16:17:05.316 27890-27890/me.ele.app.amigo W/System.err:     at me.ele.amigo.utils.component.ReceiverFinder.registerNewReceivers(Unknown Source)
10-13 16:17:05.316 27890-27890/me.ele.app.amigo W/System.err:     at me.ele.amigo.Amigo.dynamicRegisterReceivers(Unknown Source)
10-13 16:17:05.316 27890-27890/me.ele.app.amigo W/System.err:     at me.ele.amigo.Amigo.runPatchApk(Unknown Source)
10-13 16:17:05.316 27890-27890/me.ele.app.amigo W/System.err:     at me.ele.amigo.Amigo.onCreate(Unknown Source)
10-13 16:17:05.316 27890-27890/me.ele.app.amigo W/System.err:     at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1024)
10-13 16:17:05.316 27890-27890/me.ele.app.amigo W/System.err:     at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5361)
10-13 16:17:05.316 27890-27890/me.ele.app.amigo W/System.err:     at android.app.ActivityThread.-wrap2(ActivityThread.java)
10-13 16:17:05.316 27890-27890/me.ele.app.amigo W/System.err:     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1528)
10-13 16:17:05.316 27890-27890/me.ele.app.amigo W/System.err:     at android.os.Handler.dispatchMessage(Handler.java:102)
10-13 16:17:05.316 27890-27890/me.ele.app.amigo W/System.err:     at android.os.Looper.loop(Looper.java:154)
10-13 16:17:05.316 27890-27890/me.ele.app.amigo W/System.err:     at android.app.ActivityThread.main(ActivityThread.java:6077)
10-13 16:17:05.316 27890-27890/me.ele.app.amigo W/System.err:     at java.lang.reflect.Method.invoke(Native Method)
10-13 16:17:05.316 27890-27890/me.ele.app.amigo W/System.err:     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865)
10-13 16:17:05.316 27890-27890/me.ele.app.amigo W/System.err:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
10-13 16:17:05.317 27890-27890/me.ele.app.amigo W/System.err: me.ele.amigo.Amigo$LoadPatchApkException: java.lang.RuntimeException: java.lang.ClassNotFoundException: Didn't find class "me.ele.app.amigo.receiver.DemoReceiver" on path: DexPathList[[zip file "/data/app/me.ele.app.amigo-1/base.apk"],nativeLibraryDirectories=[/data/app/me.ele.app.amigo-1/lib/arm64, /system/lib64, /vendor/lib64]]
10-13 16:17:05.317 27890-27890/me.ele.app.amigo W/System.err:     at me.ele.amigo.Amigo.runPatchApk(Unknown Source)
10-13 16:17:05.317 27890-27890/me.ele.app.amigo W/System.err:     at me.ele.amigo.Amigo.onCreate(Unknown Source)
10-13 16:17:05.317 27890-27890/me.ele.app.amigo W/System.err:     at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1024)
10-13 16:17:05.318 27890-27890/me.ele.app.amigo W/System.err:     at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5361)
10-13 16:17:05.318 27890-27890/me.ele.app.amigo W/System.err:     at android.app.ActivityThread.-wrap2(ActivityThread.java)
10-13 16:17:05.318 27890-27890/me.ele.app.amigo W/System.err:     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1528)
10-13 16:17:05.318 27890-27890/me.ele.app.amigo W/System.err:     at android.os.Handler.dispatchMessage(Handler.java:102)
10-13 16:17:05.318 27890-27890/me.ele.app.amigo W/System.err:     at android.os.Looper.loop(Looper.java:154)
10-13 16:17:05.318 27890-27890/me.ele.app.amigo W/System.err:     at android.app.ActivityThread.main(ActivityThread.java:6077)
10-13 16:17:05.318 27890-27890/me.ele.app.amigo W/System.err:     at java.lang.reflect.Method.invoke(Native Method)
10-13 16:17:05.318 27890-27890/me.ele.app.amigo W/System.err:     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865)
10-13 16:17:05.318 27890-27890/me.ele.app.amigo W/System.err:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
10-13 16:17:05.318 27890-27890/me.ele.app.amigo W/System.err: Caused by: java.lang.RuntimeException: java.lang.ClassNotFoundException: Didn't find class "me.ele.app.amigo.receiver.DemoReceiver" on path: DexPathList[[zip file "/data/app/me.ele.app.amigo-1/base.apk"],nativeLibraryDirectories=[/data/app/me.ele.app.amigo-1/lib/arm64, /system/lib64, /vendor/lib64]]
10-13 16:17:05.318 27890-27890/me.ele.app.amigo W/System.err:     at me.ele.amigo.utils.component.ReceiverFinder.registerNewReceivers(Unknown Source)
10-13 16:17:05.318 27890-27890/me.ele.app.amigo W/System.err:     at me.ele.amigo.Amigo.dynamicRegisterReceivers(Unknown Source)
10-13 16:17:05.318 27890-27890/me.ele.app.amigo W/System.err: 	... 12 more
10-13 16:17:05.318 27890-27890/me.ele.app.amigo W/System.err: Caused by: java.lang.ClassNotFoundException: Didn't find class "me.ele.app.amigo.receiver.DemoReceiver" on path: DexPathList[[zip file "/data/app/me.ele.app.amigo-1/base.apk"],nativeLibraryDirectories=[/data/app/me.ele.app.amigo-1/lib/arm64, /system/lib64, /vendor/lib64]]
10-13 16:17:05.318 27890-27890/me.ele.app.amigo W/System.err:     at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
10-13 16:17:05.318 27890-27890/me.ele.app.amigo W/System.err:     at java.lang.ClassLoader.loadClass(ClassLoader.java:380)
10-13 16:17:05.318 27890-27890/me.ele.app.amigo W/System.err:     at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
10-13 16:17:05.318 27890-27890/me.ele.app.amigo W/System.err: 	... 14 more

我发现用debug版不会发现异常，如果用混淆，demo和我自己的项目就会有问题

--- a/amigo-lib/proguard-rules.txt
+++ b/amigo-lib/proguard-rules.txt
@@ -19,3 +19,5 @@
 
 -keep class me.ele.amigo.** { *; }
 -keep class **.R$** { *; }
+-keep class android.util.Singleton
+-keep class android.app.**

--- a/amigo-lib/src/main/java/me/ele/amigo/Amigo.java
+++ b/amigo-lib/src/main/java/me/ele/amigo/Amigo.java
@@ -151,7 +151,7 @@ public class Amigo extends Application {
 
             setApkInstrumentation();
             setApkHandler();
-            dynamicRegisterReceivers();
+            dynamicRegisterReceivers(amigoClassLoader);
 
             sharedPref.edit().putString(WORKING_PATCH_APK_CHECKSUM, checksum).commit();
             clearOldPatches(checksum);
@@ -188,8 +188,8 @@ public class Amigo extends Application {
         MethodUtils.invokeStaticMethod(hookFactoryClazz, "install", this, amigoClassLoader);
     }
 
-    private void dynamicRegisterReceivers() {
-        ReceiverFinder.registerNewReceivers(this);
+    private void dynamicRegisterReceivers(ClassLoader classLoader) {
+        ReceiverFinder.registerNewReceivers(this, classLoader);
     }
 
     private void setApkInstrumentation() throws Exception {

--- a/amigo-lib/src/main/java/me/ele/amigo/utils/component/ReceiverFinder.java
+++ b/amigo-lib/src/main/java/me/ele/amigo/utils/component/ReceiverFinder.java
@@ -65,13 +65,13 @@ public class ReceiverFinder extends ComponentFinder {
         return addedReceivers;
     }
 
-    public static void registerNewReceivers(Context context) {
+    public static void registerNewReceivers(Context context, ClassLoader classLoader) {
         try {
             List<ActivityInfo> addedReceivers = getNewAddedReceivers(context);
             for (ActivityInfo addedReceiver : addedReceivers) {
                 List<IntentFilter> filters = getReceiverIntentFilter(context, addedReceiver);
                 for (IntentFilter filter : filters) {
-                    BroadcastReceiver receiver = (BroadcastReceiver) ReceiverFinder.class.getClassLoader().loadClass(addedReceiver.name).newInstance();
+                    BroadcastReceiver receiver = (BroadcastReceiver) classLoader.loadClass(addedReceiver.name).newInstance();
                     context.registerReceiver(receiver, filter);
                 }
             }

--- a/app/build.gradle
+++ b/app/build.gradle
@@ -33,15 +33,17 @@ android {
         host {
             debuggable true
             signingConfig signingConfigs.debug
-            minifyEnabled false
+            minifyEnabled true
             zipAlignEnabled true
+            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
         }
 
         patch {
             debuggable true
             signingConfig signingConfigs.debug
-            minifyEnabled false
+            minifyEnabled true
             zipAlignEnabled true
+            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
         }
     }
 }

--- a/app/src/main/java/me/ele/app/amigo/DemoActivity.java
+++ b/app/src/main/java/me/ele/app/amigo/DemoActivity.java
@@ -4,12 +4,9 @@ package me.ele.app.amigo;
 import android.app.Notification;
 import android.app.NotificationManager;
 import android.app.PendingIntent;
-import android.content.ComponentName;
 import android.content.Intent;
-import android.content.ServiceConnection;
 import android.os.Bundle;
 import android.os.Environment;
-import android.os.IBinder;
 import android.support.annotation.Nullable;
 import android.support.v4.app.NotificationCompat;
 import android.support.v7.app.AppCompatActivity;
@@ -25,9 +22,6 @@ import butterknife.OnClick;
 import me.ele.amigo.Amigo;
 import me.ele.amigo.compat.RCompat;
 import me.ele.amigo.utils.CommonUtils;
-import me.ele.app.amigo.receiver.DemoReceiver;
-import me.ele.app.amigo.service.BindService;
-import me.ele.app.amigo.service.StartService;
 
 public class DemoActivity extends AppCompatActivity {
 
@@ -131,6 +125,6 @@ public class DemoActivity extends AppCompatActivity {
 
     @OnClick(R.id.trigger_receiver_action)
     public void triggerReceiverAction() {
-        sendBroadcast(new Intent("me.ele.test").putExtra(DemoReceiver.TAG, "1"));
+        sendBroadcast(new Intent("me.ele.test").putExtra("DemoReceiver", "1"));
     }
 }

--- a/app/src/patch/AndroidManifest.xml
+++ b/app/src/patch/AndroidManifest.xml
@@ -44,7 +44,7 @@
             android:label="Patched : single instance activity 2"
             android:launchMode="singleInstance"/>
 
-        <receiver android:name=".receiver.DemoReceiver">
+        <receiver android:name="me.ele.demo.receiver.DemoReceiver">
             <intent-filter>
                 <action android:name="me.ele.test"/>
             </intent-filter>

--- a/app/src/patch/java/me/ele/demo/receiver/DemoReceiver.java
+++ b/app/src/patch/java/me/ele/demo/receiver/DemoReceiver.java
@@ -1,4 +1,4 @@
-package me.ele.app.amigo.receiver;
+package me.ele.demo.receiver;
 
 import android.content.BroadcastReceiver;
 import android.content.Context;
@@ -8,7 +8,7 @@ import android.widget.Toast;
 
 public class DemoReceiver extends BroadcastReceiver {
 
-    public static final String TAG = DemoReceiver.class.getName();
+    public static final String TAG = "DemoReceiver";
 
     @Override
     public void onReceive(Context context, Intent intent) {

